File: python/basis.py
# basis
import numpy as np
from scipy.signal import windows
def basis(T,n,type):
    '''
    b=basis(T,n,type) makes basis functions of type 'hanning' or 'normal'. T
    is the length of the basis. n is how many to use. Typically T>>n, if the
    goal is to represent a filter with fewer parameters. Omit output argument
    to see how the basis functions look.
    '''
    
    r = T / n
    b = np.zeros((T, n))

    if type == 'hanning':
        for i in range(n):
            b[:int(round(r*4)), i] = windows.hann(int(round(r*4)))
        b = np.flipud(b[int(np.floor(r)):T, :])

    elif type == 'normal':
        t = np.arange(1, T+1)
        for i in range(n):
            b[:, i] = np.exp(-np.power(t - (i * r), 2) / np.power(r, 2))

    else:
        b = np.eye(T)
        
    return b

File: R/basis.R
library(pracma)  # For the hann window and matrix operations

basis <- function(T, n, type) {
  r <- T / n
  b <- matrix(0, nrow = T, ncol = n)
  
  if (type == 'hanning') {
    for (i in 1:n) {
      b[1:round(r*4), i] <- hann(round(r*4))
    }
    b <- rev(b[(floor(r) + 1):T, ])
    
  } else if (type == 'normal') {
    t <- 1:T
    for (i in 1:n) {
      b[, i] <- exp(-(t - (i * r))^2 / r^2)
    }
    
  } else {
    b <- diag(1, T, n)
  }
  
  return(b)
}